### DISCUSSION TITLE
효율성 문제 vs 이해하기 쉽고 깔끔한 코드

### DISCUSSION THREAD

**Discussion by taehyeonk** [DeveloperAcademy-POSTECH/MacC-GetARock#113] [Polls]
타임라인 화면에서 ProgressBar 처럼 좌측에 보여주는 그래프에서 선으로 연결되는 부분을 구현하고자 할 때 생긴 고민
참고할 커밋 (ab91ce6d540fb371e697f36e1dfa0d432e05f496)

해당 커밋에서 엣지 케이스로 TableView에 들어갈 Cell의 개수가 1개인 경우, 양쪽 선을 모두 없애줘야하는데
if else 구문을 쓰게 되면 둘 중 하나만 돌고 끝나기 때문에 이를 해결하기 위해 if if 로 두번 필터가 돌게 바꿨습니다.

하지만 Cell 개수가 1개일 때만 조건을 두 번 돌아야 하고 나머지는 둘 중 하나의 조건만 돌면되기 때문에 효율성이 별로라고 생각했습니다.
그래서 개수로 필터하면 되지 라고 생각했지만, 코드가 뭔가 직관적이지 않고 길어지기만 하는 것 같아서..
실제로 타임라인에 들어갈 데이터가 조건 두 개 정도를 처리하기 어려울 정도로 많아질 것 같지도 않아서 if if로 처리한 상황입니다.

예상되는 효율성 반영 코드
```
        if gathering.count == 1 {
            cell.topLine.isHidden = true
            cell.bottomLine.isHidden = true
        } else {
            if indexPath.row == 0 {
                cell.topLine.isHidden = true
            } else {
                indexPath.row == (gatheringInfos.count) - 1 {
                cell.bottomLine.isHidden = true
            }
        }
```
근데 이렇게 한다고 효율성이 높아지는게 맞나요..?
여러분들의 의견도 나누어주시면 좋을 것 같아요!

**Comment by EuniceNam** [comment]
기타) 효율성에 안 좋을 수 있지만 비슷한 문제를 다르게 했던 방법 남깁니다. 가독성 위주입니다. 
* 테이블뷰의 첫 셀의 위와 마지막 셀의 아래 구분선을 감추는 코드입니다.
* (디자인을 잘못 생각한 것이었어서 해당 커밋은 삭제함)

Cell에 이런 메소드를 추가하고
```swift
public func setupSeparator(cellAtRow: Int, totalRowNumber: Int) {
    upperSeparatorView.isHidden = (cellAtRow == 0)
    lowerSeparatorView.isHidden = (cellAtRow == totalRowNumber - 1)
}
```
DataSource에서 이렇게 썼습니다.
```swift
cell.setupSeparator(cellAtRow: (indexPath as NSIndexPath).row, totalRowNumber: 어레이.count)
```
문제의 차이점
* 이 셀을 2개 데이터소스에서 써서 Cell에 메소드를 넣은 면도 있습니다. 
* if문 대신 `cellAtRow == 0`처럼 쓴 이유는 다른 상황에서 `isHidden = false`도 해줘야 셀의 결과가 이상하게 나오지 않았기 때문입니다.
